cmd/endpoint.go

```
@@ -0,0 +1,382 @@
+		if err != nil {
+			return serverAddr, endpoints, setupType, err
+		}
+
+		if endpoint.Type() != PathEndpointType {
+			return serverAddr, endpoints, setupType, fmt.Errorf("use path style endpoint for FS setup")
+		}
+
+		endpoints = append(endpoints, endpoint)
+		setupType = FSSetupType
+		return serverAddr, endpoints, setupType, nil
+	}
+
+	// Convert args to endpoints
+	if endpoints, err = NewEndpointList(args...); err != nil {
+		return serverAddr, endpoints, setupType, err
+	}
+
+	// Return XL setup when all endpoints are path style.
+	if endpoints[0].Type() == PathEndpointType {
+		setupType = XLSetupType
+		return serverAddr, endpoints, setupType, nil
+	}
+
+	// Here all endpoints are URL style.
+	endpointPathSet := set.NewStringSet()
+	localEndpointCount := 0
+	localServerAddrSet := set.NewStringSet()
+	localPortSet := set.NewStringSet()
+	for _, endpoint := range endpoints {
+		endpointPathSet.Add(endpoint.Path)
+		if endpoint.IsLocal {
+			localServerAddrSet.Add(endpoint.Host)
+
+			var port string
+			_, port, err = net.SplitHostPort(endpoint.Host)
+			if err != nil {
+				port = serverAddrPort
+			}
+
+			localPortSet.Add(port)
+
+			localEndpointCount++
+		}
+	}
+
+	// No local endpoint found.
+	if localEndpointCount == 0 {
+		return serverAddr, endpoints, setupType, fmt.Errorf("no endpoint found for this host")
+	}
+
+	// Check whether same path is not used in endpoints of a host.
+	{
+		pathIPMap := make(map[string]set.StringSet)
+		for _, endpoint := range endpoints {
+			var host string
+			host, _, err = net.SplitHostPort(endpoint.Host)
+			if err != nil {
+				host = endpoint.Host
+			}
+			hostIPSet, _ := getHostIP4(host)
+			if IPSet, ok := pathIPMap[endpoint.Path]; ok {
+				if !IPSet.Intersection(hostIPSet).IsEmpty() {
+					err = fmt.Errorf("path '%s' can not be served from different address/port", endpoint.Path)
+					return serverAddr, endpoints, setupType, err
+				}
+			} else {
+				pathIPMap[endpoint.Path] = hostIPSet
+			}
+		}
```

In the above, the algorithm seems incorrect, and the error message is also misleading.

It is legitimate to serve a path like `/export` from different hosts. It is not legitimate to serve `/export` from multiple minio servers on the same host. The error message is conveying the opposite of this.

The algorithm is incorrect - e.g. endpoints:

```
10.0.0.1:9000/export
10.0.0.2:9000/export
10.0.0.2:9001/export
10.0.0.3:9000/export
```

After first iteration, the pathIPMap will have an entry for `/export` with stringset value containing just the IP `10.0.0.1`. In the second iteration, the IP seen is `10.0.0.2` and no error is raised. In the third iteration, the same IP 10.0.0.2 is seen - this should return an error, but this will also pass undetected because the intersection operation will give an empty set.

---

> In the above, the algorithm seems incorrect, and the error message is also misleading.

What is the right error message you are expecting?



> It is legitimate to serve a path like /export from different hosts. It is not legitimate to serve /export from multiple minio servers on the same host. The error message is conveying the opposite of this.
> 
> The algorithm is incorrect - e.g. endpoints:
> 
> 10.0.0.1:9000/export
> 10.0.0.2:9000/export
> 10.0.0.2:9001/export
> 10.0.0.3:9000/export
> 
> After first iteration, the pathIPMap will have an entry for /export with stringset value containing just the IP 10.0.0.1. In the second iteration, the IP seen is 10.0.0.2 and no error is raised. In the third iteration, the same IP 10.0.0.2 is seen - this should return an error, but this will also pass undetected because the intersection operation will give an empty set.


It seems you lost something there.

Below test case passes
```go
	case6u1, _ := url.Parse("http://10.0.0.1:9000/export")
	case6u2, _ := url.Parse("http://10.0.0.2:9000/export")
	case6u3, _ := url.Parse("http://10.0.0.3:9000/export")
	case6u4, _ := url.Parse("http://localhost:9001/export")

		{":9001", []string{"http://10.0.0.1:9000/export", "http://10.0.0.2:9000/export", "http://localhost:9001/export", "http://10.0.0.3:9000/export"}, ":9001", EndpointList{
			Endpoint{URL: case6u1, IsLocal: false},
			Endpoint{URL: case6u2, IsLocal: false},
			Endpoint{URL: case6u3, IsLocal: false},
			Endpoint{URL: case6u4, IsLocal: true},
		}, DistXLSetupType, nil},
```

---

Please try the test case I provided rather than the above. The bug is in the test case that I provided.

---

@donatello I got the problem and fixed in https://github.com/minio/minio/pull/4105